File: ABC273/d-review.cpp
#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <vector>

using i64 = std::int64_t;

int main() {
    size_t h, w, n;
    i64 r, c;
    std::cin >> h >> w >> r >> c >> n;
    std::vector<i64> walls_rc(n, 0), walls_cr(n, 0);
    for(size_t i = 0; i < n; ++i) {
        i64 r, c;
        std::cin >> r >> c;
        r--;
        c--;
        walls_rc[i] = r + c * h;
        walls_cr[i] = c + r * w;
    }
    std::sort(walls_rc.begin(), walls_rc.end());
    std::sort(walls_cr.begin(), walls_cr.end());
    r--;
    c--;
    size_t q;
    std::cin >> q;
    for(size_t _ = 0; _ < q; ++_) {
        char d;
        i64 l;
        std::cin >> d >> l;
        if(d == 'U') {
            auto i =
                std::lower_bound(walls_rc.begin(), walls_rc.end(), r + c * h);
            i64 border = 0;
            if(i == walls_rc.begin() || *(i - 1) / h != c) {
                border = 0;
            } else {
                border = (*(i - 1) % h) + 1;
            }
            r = std::max(std::max(static_cast<i64>(0), r - l), border);
        } else if(d == 'D') {
            auto i =
                std::lower_bound(walls_rc.begin(), walls_rc.end(), r + c * h);
            i64 border = 0;
            if(i == walls_rc.end() || *i / h != c) {
                border = h;
            } else {
                border = *i % h;
            }
            r = std::min(std::min(static_cast<i64>(h - 1), r + l), border - 1);
        } else if(d == 'L') {
            auto i =
                std::lower_bound(walls_cr.begin(), walls_cr.end(), c + r * w);
            i64 border = 0;
            if(i == walls_cr.begin() || *(i - 1) / w != r) {
                border = 0;
            } else {
                border = (*(i - 1) % w) + 1;
            }
            c = std::max(std::max(static_cast<i64>(0), c - l), border);
        } else if(d == 'R') {
            auto i =
                std::lower_bound(walls_cr.begin(), walls_cr.end(), c + r * w);
            i64 border = 0;
            if(i == walls_cr.end() || *i / w != r) {
                border = w;
            } else {
                border = *i % w;
            }
            c = std::min(std::min(static_cast<i64>(w - 1), c + l), border - 1);
        } else {
            abort();
        }
        std::cout << (r + 1) << " " << (c + 1) << std::endl;
    }
}

File: ABC273/d-review.rs
use std::cmp;

/// vにxを挿入しても順序が保たれる最小の位置を返す
fn lower_bound(v: &[i64], x: i64) -> usize {
    let mut ng = -1;
    let mut ok = v.len() as isize;
    while ok - ng > 1 {
        let mid = (ok + ng) / 2;
        if v[mid as usize] >= x {
            ok = mid;
        } else {
            ng = mid;
        }
    }
    ok as usize
}

fn main() {
    proconio::input! {
        h: i64, w: i64, r: i64, c: i64,
        n: usize,
        walls: [(i64, i64); n],
        q: usize,
        queries: [(char, i64); q],
    }
    let mut walls_rc = vec![0; n];
    let mut walls_cr = vec![0; n];
    for (i, &(r, c)) in walls.iter().enumerate() {
        let r = r - 1;
        let c = c - 1;
        walls_rc[i] = r + c * h;
        walls_cr[i] = c + r * w;
    }
    walls_rc.sort();
    walls_cr.sort();
    let mut r = r - 1;
    let mut c = c - 1;
    for (d, l) in queries {
        match d {
            'U' => {
                let i = lower_bound(&walls_rc, r + c * h);
                let border = if i == 0 || walls_rc[i - 1] / h != c {
                    0
                } else {
                    (walls_rc[i - 1] % h) + 1
                };
                r = cmp::max(cmp::max(0, r - l), border);
            }
            'D' => {
                let i = lower_bound(&walls_rc, r + c * h);
                let border = if i == walls_rc.len() || walls_rc[i] / h != c {
                    h - 1
                } else {
                    (walls_rc[i] % h) - 1
                };
                r = cmp::min(cmp::min(h - 1, r + l), border);
            }
            'L' => {
                let i = lower_bound(&walls_cr, c + r * w);
                let border = if i == 0 || walls_cr[i - 1] / w != r {
                    0
                } else {
                    walls_cr[i - 1] % w + 1
                };
                c = cmp::max(cmp::max(0, c - l), border);
            }
            'R' => {
                let i = lower_bound(&walls_cr, c + r * w);
                let border = if i == walls_cr.len() || walls_cr[i] / w != r {
                    w - 1
                } else {
                    walls_cr[i] % w - 1
                };
                c = cmp::min(cmp::min(w - 1, c + l), border);
            }
            _ => unreachable!(),
        }
        println!("{} {}", r + 1, c + 1);
    }
}
